### PR TITLE
clojure: Pass function to mkDerivation, remove rec

### PR DESCRIPTION
**Pull request by jlesquembre**
See
https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes

It also it updates clojure to latest version

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Please split the version bump and the refactor in 2 separate commits.

**Comment by jlesquembre** [comment]
@thiagokokada done

File: pkgs/development/interpreters/clojure/default.nix [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
   pname = "clojure";
-  version = "1.11.1.1273";
+  version = "1.11.1.1342";
 
   src = fetchurl {
     # https://clojure.org/releases/tools
-    url = "https://download.clojure.org/install/clojure-tools-${version}.tar.gz";
-    sha256 = "sha256-X4uvzyS9FIrJvL5gqOe4CTye2OuODzhxmXcOOPDkDOY=";
+    url = "https://download.clojure.org/install/clojure-tools-${finalAttrs.version}.tar.gz";
+    sha256 = "sha256-obxgp7etdCI+jikoGcppmI2pyIdH6In2n870wQJlRY4=";
```
Comment (drupol) [inline pkgs/development/interpreters/clojure/default.nix:10]: Please use `hash` instead of `sha256`.

**Review by drupol — CHANGES_REQUESTED** [review]
One minor nit and it's good to go.

Thanks for implementing the `finalAttrs` pattern.

**Review by drupol — CHANGES_REQUESTED** [review]
May I ask you to have 2 commits in the end ?
1. One for the `finalAttrs` pattern
2. One for the derivation version bump

File: pkgs/development/interpreters/clojure/default.nix (hunk 1 above)
Comment (jlesquembre) [reply pkgs/development/interpreters/clojure/default.nix]: @drupol Done :)

**Comment by jlesquembre** [comment]
@drupol There are 2 commits, isn't it? Or am I missing something?

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Sorry my bad, everything's ok now, let's wait the eval and then we're good to go.

**Pull request merged**